benchmarks/data_utils.py

```diff
@@ -0,0 +1,119 @@
+import os
+
+import torch
+from torch.utils.data import Dataset
+
+import torchvision
+import torchvision.transforms.functional as TF
+
+
+
+
+###
+# The dataset initializes with the file path to frame pairs
+# and loads the image when get image is called. Did this back when planning to take frames 
+# from entire video. Now the data is stored as as only 2 frames, so not really a necessary step
+###
+class FramePairDataset(Dataset):
+    def __init__(self, path_to_folder, class_types=['control', 'mdivi', 'llo'], transform=None):
+        self.targets= []
+        self.vid_path = []
+        self.transform = transform
+        # self.transform_input = transform_input
+        self.class_types = class_types
+
+        print(path_to_folder)
+        
+
+        name_constraint = lambda x: 'normalized' in x
+        for label in self.class_types:
+            target =  self.class_types.index(label) #class 0,1,2
+            path = os.path.join(path_to_folder, label)
+            files = os.listdir(path)
+            for file_name in files:
+                    if  name_constraint(file_name):
+                        self.vid_path.append(os.path.join(path, file_name))
+                        self.targets.append(target)
+        
+    def __len__(self):
+        return len(self.targets)
+
+    
+    def __getitem__(self, idx):
+        target_class = self.targets[idx]
+        vid = np.load(self.vid_path[idx])
+        first_frame = vid[0]
+        last_frame = vid[-1] 
+
+        first_frame = torch.as_tensor((first_frame))
+        last_frame = torch.as_tensor((last_frame))
+
+        # stack frames as channels, x shape is [2, img_width, img_height]
+        sample = torch.stack((first_frame, last_frame))
+
+       
+        if self.transform:
+            sample = self.transform(sample)
+            
+        return sample, target_class
+
+
+#TODO is this batchable, I think it is
+# TODO, return some indication that image is empty. 
+    #TODO - if it ends up empty, did protein move or dissapear?? how much do we handle at the moment
+#TODO also this is kinda slow if that becomes a problem.
+class RoiTransform:
+    """Pick a ROI based on pixel intensity. Choose the same coordinates for both/all frames"""
+
+    def __init__(self, window_size=(28,28), kernel_size=5, use_first_frame_values=True):
+        """
+        window_size = (width, height) of cropped section  
+        kernel_size = size of gaussian kernel used in selecting peak pixel values  
+        use_first_frame_values = if true, set ROI based off first frame, if false, set ROI based off last frame
+        """
+        self.width = window_size[0]
+        self.height = window_size[1]
+        self.kernel_size = kernel_size
+        self.use_first_frame = use_first_frame_values
+
+    def __call__(self, sample):
+        """
+        sample is [n_channels, img_width, img_height]
+        """
+
+        first_frame, last_frame = sample.split(1, dim=0)
```

Comment:
I'm not sure why this would work -- if the incoming data is of shape `[n_channels, img_width, img_height]` then it seems that it's coming in _after_ a conv layer, in which case we can't differentiate any channel as the "first" or "last" frames. 

Instead, perhaps it should be the case that the expected shape is just `[2, img_width, img_height]`? In which case it should suffice to set 
```python
first_frame, last_frame = sample
```

Reply:
Yeah, I wanted the results to be [1, width, height]. I think one of the subsequent torch functions needs that format.